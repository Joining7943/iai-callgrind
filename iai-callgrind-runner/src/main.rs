use std::io::Write;

use colored::{control, Colorize};
use env_logger::Env;
use iai_callgrind_runner::error::IaiCallgrindError;
use iai_callgrind_runner::runner::envs;
use iai_callgrind_runner::util::write_all_to_stderr;
use log::{error, warn};
use version_compare::Cmp;

fn print_warnings() {
    if std::env::var("IAI_ALLOW_ASLR").is_ok() {
        warn!("The IAI_ALLOW_ASLR environment variable changed to IAI_CALLGRIND_ALLOW_ASLR");
    }

    if std::env::var("RUST_LOG").is_ok() {
        warn!(
            "The RUST_LOG environment variable to set the log level changed to IAI_CALLGRIND_LOG"
        );
    }
}

fn main() {
    // Configure the colored crate to respect IAI_CALLGRIND_COLOR and CARGO_TERM_COLOR
    let iai_callgrind_color = std::env::var(envs::IAI_CALLGRIND_COLOR).ok();
    if let Some(var) = iai_callgrind_color
        .as_ref()
        .or(std::env::var(envs::CARGO_TERM_COLOR).ok().as_ref())
    {
        if var == "never" {
            control::set_override(false);
        } else if var == "always" {
            control::set_override(true);
        }
    }

    // Configure the env_logger crate to respect IAI_CALLGRIND_COLOR and CARGO_TERM_COLOR
    env_logger::Builder::from_env(
        Env::default()
            .filter_or(envs::IAI_CALLGRIND_LOG, "warn")
            .write_style(
                iai_callgrind_color
                    .map_or_else(|| envs::CARGO_TERM_COLOR, |_| envs::IAI_CALLGRIND_COLOR),
            ),
    )
    .format(|buf, record| {
        writeln!(
            buf,
            "{}: {:<5}: {}",
            record
                .module_path()
                .unwrap_or(record.module_path_static().unwrap_or("???")),
            match record.level() {
                log::Level::Error => "Error".red().bold(),
                log::Level::Warn => "Warn".yellow().bold(),
                log::Level::Info => "Info".green().bold(),
                log::Level::Debug => "Debug".blue().bold(),
                log::Level::Trace => "Trace".cyan().bold(),
            },
            record.args()
        )
    })
    .init();

    print_warnings();
    match iai_callgrind_runner::runner::run() {
        Ok(_) => {}
        Err(error) => {
            match error {
                IaiCallgrindError::VersionMismatch(cmp, runner_version, library_version) => {
                    match cmp {
                        Cmp::Lt => error!(
                            "iai-callgrind-runner ({0}) is older than iai-callgrind ({1}). Please \
                             update iai-callgrind-runner by calling 'cargo install --version {1} \
                             iai-callgrind-runner'",
                            runner_version, library_version
                        ),
                        Cmp::Gt => error!(
                            "iai-callgrind-runner ({0}) is newer than iai-callgrind ({1}). Please \
                             update iai-callgrind to '{0}' in your Cargo.toml file",
                            runner_version, library_version
                        ),
                        Cmp::Ne => error!(
                            "No version information found for iai-callgrind but \
                             iai-callgrind-runner ({0}) is >= '0.3.0'. Please update \
                             iai-callgrind to '{0}' in your Cargo.toml file",
                            runner_version
                        ),
                        _ => unreachable!(),
                    }
                }
                IaiCallgrindError::LaunchError(exec, error) => {
                    error!("Error executing '{}': {}", exec.display(), error)
                }
                IaiCallgrindError::BenchmarkLaunchError(output) => {
                    error!("Captured stderr:",);
                    write_all_to_stderr(&output.stderr);
                    error!(
                        "Error launching benchmark: Exit code was: {}",
                        output.status.code().unwrap()
                    );
                }
                IaiCallgrindError::Other(message) => {
                    error!("{}", message);
                }
                IaiCallgrindError::InvalidCallgrindBoolArgument((option, value)) => {
                    error!(
                        "Invalid callgrind argument for --{option}: '{value}'. Valid values are \
                         'yes' or 'no'"
                    );
                }
            }
            std::process::exit(1)
        }
    }
}
