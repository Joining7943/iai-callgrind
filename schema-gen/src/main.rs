//! A utility binary to create the json schema for the summary.json file
//!
//! This binary is not considered a part of the published `iai-callgrind-runner` package and is only
//! used during the development of `iai-callgrind`.
use std::fs::File;

use iai_callgrind_runner::runner::summary::BenchmarkSummary;
use schemars::schema_for;

fn main() {
    serde_json::to_writer_pretty(
        File::create("summary.schema.json").unwrap(),
        &schema_for!(BenchmarkSummary),
    )
    .expect("Schema creation should be successful");
}
