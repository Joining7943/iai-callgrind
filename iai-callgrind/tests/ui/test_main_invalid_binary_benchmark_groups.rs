mod test_main_when_binary_benchmark_group_is_not_a_group {
    use iai_callgrind::main;

    fn some_func() {}

    main!(binary_benchmark_groups = some_func);
}

mod test_main_when_config_is_not_a_binary_benchmark_config {
    use iai_callgrind::{binary_benchmark_group, main};

    binary_benchmark_group!(
        name = some;
        benchmarks = |group: &mut BinaryBenchmarkGroup| {
            // do nothing
        }
    );

    main!(
        config = "some string";
        binary_benchmark_groups = some
    );
}

mod test_main_when_no_group {
    use iai_callgrind::main;
    main!(binary_benchmark_groups = );
}

fn main() {}
