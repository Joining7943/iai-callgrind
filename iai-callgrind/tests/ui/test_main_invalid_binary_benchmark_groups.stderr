error: The binary_benchmark_groups argument needs at least one `name` of a `binary_benchmark_group!`
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:27:5
   |
27 |     main!(binary_benchmark_groups = );
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: function `some_func` is not a crate or module
 --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:6:37
  |
6 |     main!(binary_benchmark_groups = some_func);
  |                                     ^^^^^^^^^ function `some_func` is not a crate or module

error[E0277]: the trait bound `InternalBinaryBenchmarkConfig: From<&str>` is not satisfied
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:19:5
   |
19 | /     main!(
20 | |         config = "some string";
21 | |         binary_benchmark_groups = some
22 | |     );
   | |_____^ the trait `From<&str>` is not implemented for `InternalBinaryBenchmarkConfig`, which is required by `&str: Into<_>`
   |
   = help: the following other types implement trait `From<T>`:
             `InternalBinaryBenchmarkConfig` implements `From<&BinaryBenchmarkConfig>`
             `InternalBinaryBenchmarkConfig` implements `From<&mut BinaryBenchmarkConfig>`
             `InternalBinaryBenchmarkConfig` implements `From<BinaryBenchmarkConfig>`
   = note: required for `&str` to implement `Into<InternalBinaryBenchmarkConfig>`
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused variable: `group`
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:14:23
   |
14 |         benchmarks = |group: &mut BinaryBenchmarkGroup| {
   |                       ^^^^^
   |
help: `group` is captured in macro and introduced a unused variable
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:12:5
   |
12 | /     binary_benchmark_group!(
13 | |         name = some;
14 | |         benchmarks = |group: &mut BinaryBenchmarkGroup| {
15 | |             // do nothing
16 | |         }
17 | |     );
   | |_____^
   = note: `#[warn(unused_variables)]` on by default
   = note: this warning originates in the macro `binary_benchmark_group` (in Nightly builds, run with -Z macro-backtrace for more info)
